### ISSUE TITLE
Disk not recognized

### ISSUE THREAD

**Issue by GusJeep** [lwouis/fat-drive-sorter#1]
**Describe the bug**

I have a 256 GB HDD Fat 32 formated and it is not recognized.

**Your environment**

* FatDriveSorter version: 1.1.0
* macOS version: Ventura 13.5

**Comment by lwouis** [comment]
Hi,

Could you please share more details?

Is it recognized by macOS? Can you read/write contents?
What does FatDriveSorter show?

Thank you

**Comment by GusJeep** [comment]
Hi!!
Yes, it is recognized by mac os and in my car. I tried with a pen drive and it recognize it well, but the hdd doesn’t.
 I use it for playing music.

Thanks!

Gus Battisti

**Comment by lwouis** [comment]
Could you share a screenshot of how FatDriveSorter main window shows?

**Comment by GusJeep** [comment]
Yes, sure!
￼
Gus Battisti

**Comment by lwouis** [comment]
I'm sorry, I can't see the image

**Comment by GusJeep** [comment]
Sorry! It was png. I send you a jpeg.
And a lot of thanks for your effort!!!￼
Gus Battisti

**Comment by lwouis** [comment]
I'm sorry, I still can't see it. I think it's the way you send the messages through email. I suggest using github in the web browser

**Comment by GusJeep** [comment]
I Try Pdf.
￼
Gus Battisti

**Comment by lwouis** [comment]
<img width="740" alt="image" src="https://github.com/lwouis/fat-drive-sorter/assets/106195/268a0900-b244-4c4c-9049-8497b067bd6a">

**Comment by GusJeep** [comment]
Gus Battisti

**Comment by GusJeep** [comment]
Hi! I send you the screenshot.
![Captura de pantalla 2023-08-31 a la(s) 19 27 01](https://github.com/lwouis/fat-drive-sorter/assets/143756528/6eb5cdd6-0d13-46a0-9205-7d5d1d294860)

**Comment by lwouis** [comment]
Thank you for sharing this screenshot.

Here's [the code which lists mounted drives](https://github.com/lwouis/fat-drive-sorter/blob/af95bf0761157cf288b5a2fda90f73346a82623b/src/ui/Window.swift#L54).

FatDriveSorter asks the OS for the drives, straight up. It's surprising that the OS is not listing the drive here, since you mentioned that you see it in Finder.

The only thing I can imagine is that somehow your drive is not removable or ejectable. FatDriveSorter will only sort removable and ejectable drives. Maybe it was a mistake to not show non-removable drives. I added that check because in my tests, all external drives had that. But I found this definition online which seems to imply that we shouldn't check those probably:

> As they use the term, "removable" means that the media and the drive for the media are separate (like a floppy drive or CD-ROM). Anything that doesn't allow a single drive to hold different media at different times is a "fixed" drive.

I've created a local build which doesn't have that check. Could you please test [this build](https://github.com/lwouis/fat-drive-sorter/files/12499297/FatDriveSorter.app.zip) and let me know if it lists your drive?

If it works for you, I'll release an official version with this fix.

Thank you

**Comment by GusJeep** [comment]
Hi! Sorry for the inconvenients! It doesn´t let me open it.
<img width="497" alt="Captura de pantalla 2023-09-01 a la(s) 13 13 13" src="https://github.com/lwouis/fat-drive-sorter/assets/143756528/66cc50a5-7590-4ad0-94d4-60a973def86e">

**Comment by GusJeep** [comment]
Finally I could open FatDriveSorter but it still not detect de HDD. Sorry! And thanks a lot for your effort!

**Comment by GusJeep** [comment]
The Disk is a SSD 240 GB, formated MS-DOS Fat 32.

**Comment by lwouis** [comment]
I'm surprised it didn't work. I wonder why it's not showing. Is the drive not ejectable somehow? Can you eject it?

Could you please open `Disk Utility.app` and screenshot the drive information? Usually there are multiple nested sections within a drive. Could you please share each?

**Comment by GusJeep** [comment]
Hi!! I´m really sorry for bothering you for this inconvenient and I appreciate so much your effort!
Yes, I can eject the disk normally. I also tried with another SSD and another case and it doesn´t detect them.
This are the
<img width="1183" alt="Captura de pantalla 2023-09-01 a la(s) 18 43 55" src="https://github.com/lwouis/fat-drive-sorter/assets/143756528/0a002207-0830-4455-bfd2-4da8f5b4bdcd">
 screenshots:
<img width="1280" alt="Captura de pantalla 2023-09-01 a la(s) 18 45 12" src="https://github.com/lwouis/fat-drive-sorter/assets/143756528/2dacc373-e346-46fb-a8b1-f70c05b6ad5c">

**Comment by lwouis** [comment]
Thank you for sharing these screenshots.

The Drive looks healthy. It's intriguing why the app is not listing it.

Could you please try [this build](https://github.com/lwouis/fat-drive-sorter/files/12501748/FatDriveSorter.app.zip)?

**Comment by GusJeep** [comment]
Yes! Now it detects the disk, but I can not see the directory order, when I click on see current file order, the drive tab changes to Mac Disk.

**Comment by GusJeep** [comment]
Can it be something with mbr?

**Comment by lwouis** [comment]
Sorry to ask again: could you please try and [this build](https://github.com/lwouis/fat-drive-sorter/files/12501901/FatDriveSorter.app.zip) and share a screenshot of the list of Drives shown in the app?

**Comment by GusJeep** [comment]
Please don´t you sorry! No, didn´t work. Maybe is the way my SSD is formated. With a pendrive the original version works well.

**Comment by lwouis** [comment]
Hi @GusJeep,

Since I can't reproduce the issue on my side, I need your help to investigate. Could you please run [this build](https://github.com/lwouis/fat-drive-sorter/files/12516339/FatDriveSorter.app.zip) from Terminal.app?

In order to do this, here are steps to guide you:

* Download the build I linked above. Unzip it. Move `FatDriveSorter.app` to the `/Applications` folder
* Open Terminal.app. You can find it in `/Applications/Utilities`
* On the window, paste this command: `/Applications/FatDriveSorter.app/Contents/MacOS/FatDriveSorter`
* Press enter
* It should now launch FatDriveSorter, and you should see logs in the Terminal window
* Please share those logs here, so I may be able to deduct what's happening on your side

Thank you

**Comment by GusJeep** [comment]
Hii!! Again, sorry for all this extra work! Here I send you the sanpshot. The disk JeepBup is the Fat32 one. Thanks !!
<img width="613" alt="Captura de pantalla 2023-09-04 a la(s) 17 31 54" src="https://github.com/lwouis/fat-drive-sorter/assets/143756528/46a86755-050f-4b35-bb1c-66b6eeeb5e5d">

**Comment by lwouis** [comment]
Thank you for sharing this screenshot. So it's confirmed: this disk is not showing because it's flagged by macOS as "not ejectable".

As I said above, when I wrote this app, I thought we should only list external drives, not internal drives. I don't see a use-case for wanting to alphabetize an internal drive. People want to alphabetize external drives which they use with mp3 players, and other devices.

There is technical way to filter only "external drive" with the macOS APIs. So I used "ejectable" as a proxy for "external".

I'm surprised to see you have an external drive here which is deemed "not ejectable" by macOS. I'm not sure how that makes sense. And I'm not sure how I could exclude internal drives without using the "is ejectable?" check.

**Comment by GusJeep** [comment]
Hi! I use this disk for playing mp3 music and video music in my car. I don´t have any problem ejecting the disk from my Mac. I don´t know how can I help you to fix it, or there is no fix? Thanks!

**Comment by lwouis** [comment]
Could you please try [this build](https://github.com/lwouis/fat-drive-sorter/files/12516782/FatDriveSorter.app.zip)? You can launch it normally.

I removed the check on "is the drive ejectable". To mitigate risk for the users to format their internal drives, I added a check to only show "FAT" drives (e.g. FAT16, FAT32, exFAT).

Could you please tell me if it lists your drive, or not other drives?

Thank you

**Comment by GusJeep** [comment]
Now I can see the disk, but it doesn´t show directories or the see current file order doesn´t show anything

**Comment by lwouis** [comment]
Do you mean that you clicked "See current file order", and it didn't show you anything? Could you please share a screenshot?

**Comment by GusJeep** [comment]
https://github.com/lwouis/fat-drive-sorter/assets/143756528/e4c8a5a4-c0a9-40a6-ac04-a71e61d04de1

**Comment by lwouis** [comment]
Oh right, I didn't think of this.

The way the app works is that when you do actions like listing files, or sorting the drive, it will first unmount the drive, then do the operation on the unmounted drive, then remount it so it's accessible to the user again.

It's a technical requirement to operate on the unmounted drive. We can't list or sort files on a mounted drive.

I guess your drive not being ejectable means it can't be unmounted?

I wish I had a drive like yours at home to play with and understand why it's not ejectable. I don't know what can be done with such a drive. I don't see why it can't be ejected or unmounted since it's an external USB drive.

I'm sorry, I'm running out of ideas here to get it to work on the drive 🙇

**Comment by GusJeep** [comment]
Ok, please don´t be sorry! I really appreciate all you have done to make it work! Thank you a lot! Don´t worry about this, Mac Os isn´t easy! Best regards!!!

**Comment by lwouis** [comment]
Could you please run the latest build I shared above, from the Terminal?

I think it will log interesting things when you press "See current file order". Could you please share those logs?

**Comment by GusJeep** [comment]
Hey!!!! Now it works!!!! Great job! Thank you so much!! You are a genius!!
